File: ext/group_reply.cpp
/******************************************************************************
  This file is part of PyTango (http://pytango.rtfd.io)

  Copyright 2006-2012 CELLS / ALBA Synchrotron, Bellaterra, Spain
  Copyright 2013-2019 European Synchrotron Radiation Facility, Grenoble, France

  Distributed under the terms of the GNU Lesser General Public License,
  either version 3 of the License, or (at your option) any later version.
  See LICENSE.txt for more info.
******************************************************************************/

#include <tango.h>
#include <pybind11/pybind11.h>
#include <defs.h>
#include "device_attribute.h"

namespace py = pybind11;

void export_group_reply(py::module &m) {
    py::class_<Tango::GroupReply>(m, "GroupReply")
        .def("has_failed", [](Tango::GroupReply& self) -> bool {
            return self.has_failed(); // Tango C++ signature
        })
        .def("group_element_enabled", [](Tango::GroupReply& self) -> bool {
            return self.group_element_enabled(); // Tango C++ signature
        })
        .def("dev_name", [](Tango::GroupReply& self) -> std::string {
            std::string ret = static_cast<std::string>(self.dev_name()); // Tango C++ signature
            return ret;
        })
        .def("obj_name", [](Tango::GroupReply& self) -> std::string {
            std::string ret = static_cast<std::string>(self.obj_name()); // Tango C++ signature
            return ret;
        })
        .def("get_err_stack", [](Tango::GroupReply& self) -> Tango::DevErrorList {
            return self.get_err_stack(); // Tango C++ signature
        })
    ;

    py::class_<Tango::GroupCmdReply, Tango::GroupReply>(m, "GroupCmdReply")
        .def("get_data_raw", [](Tango::GroupCmdReply& self) -> Tango::DeviceData& {
            return self.get_data(); // Tango C++ signature
        })
    ;

    py::class_<Tango::GroupAttrReply, Tango::GroupReply>(m, "GroupAttrReply")
        .def("__get_data", [](Tango::GroupAttrReply& self) -> py::object {
                // Usually we pass a device_proxy to "convert_to_python" in order to
                // get the data_format of the DeviceAttribute for Tango versions
                // older than 7.0. However, GroupAttrReply has no device_proxy to use!
                // So, we are using update_data_format() in:
                // GroupElement::read_attribute_reply/read_attributes_reply
            return PyDeviceAttribute::convert_to_python(
                new Tango::DeviceAttribute(self.get_data())); // Tango C++ signature
        })
    ;
}

File: tango/group_reply.py
# ------------------------------------------------------------------------------
# This file is part of PyTango (http://pytango.rtfd.io)
#
# Copyright 2006-2012 CELLS / ALBA Synchrotron, Bellaterra, Spain
# Copyright 2013-2014 European Synchrotron Radiation Facility, Grenoble, France
#
# Distributed under the terms of the GNU Lesser General Public License,
# either version 3 of the License, or (at your option) any later version.
# See LICENSE.txt for more info.
# ------------------------------------------------------------------------------

"""
This is an internal PyTango module.
"""

__all__ = ("group_reply_init",)

__docformat__ = "restructuredtext"

from .utils import document_method as __document_method
from ._tango import GroupReply, GroupCmdReply, GroupAttrReply, ExtractAs


def __GroupCmdReply__get_data(self):
    return self.get_data_raw().extract()


def __GroupAttrReply__get_data(self, extract_as=ExtractAs.Numpy):
    # GroupAttrReply.__get_data() extracts the data from the object, so
    # two successive calls to get_data() result in the second one returning
    # an empty value, which is an unexpected behaviour.
    # That's why we cache the result of the first call.
    try:
        data, orig_extract_as = self.__dataCache
    except AttributeError:
        data = self.__get_data(extract_as)
        self.__dataCache = data, extract_as
        return data

    if extract_as != orig_extract_as:
        raise Exception("Successive calls to get_data() must receive the same"
                        " parameters as the first one.")
    return data


def __init_GroupReply():
    GroupCmdReply.get_data = __GroupCmdReply__get_data
    GroupAttrReply.get_data = __GroupAttrReply__get_data


def __doc_GroupReply():
    def document_method(method_name, desc, append=True):
        return __document_method(GroupReply, method_name, desc, append)

    GroupReply.__doc__ = """
        This is the base class for the result of an operation on a
        PyTangoGroup, being it a write attribute, read attribute, or
        command inout operation.

        It has some trivial common operations:

            - has_failed(self) -> bool
            - group_element_enabled(self) ->bool
            - dev_name(self) -> str
            - obj_name(self) -> str
            - get_err_stack(self) -> DevErrorList
    """

    __document_method(GroupCmdReply, "get_data", """
    get_data(self) -> any

            Get the actual value stored in the GroupCmdRply, the command
            output value.
            It's the same as self.get_data_raw().extract()

        Parameters : None
        Return     : (any) Whatever is stored there, or None.
    """)

    __document_method(GroupCmdReply, "get_data_raw", """
    get_data_raw(self) -> any

            Get the DeviceData containing the output parameter
            of the command.

        Parameters : None
        Return     : (DeviceData) Whatever is stored there, or None.
    """)

    __document_method(GroupAttrReply, "get_data", """
    get_data(self, extract_as=ExtractAs.Numpy) -> DeviceAttribute

            Get the DeviceAttribute.

        Parameters :
            - extract_as : (ExtractAs)

        Return     : (DeviceAttribute) Whatever is stored there, or None.
    """)


def group_reply_init(doc=True):
    __init_GroupReply()
    if doc:
        __doc_GroupReply()
